pkg/querier/queryrange/querysharding.go

```diff
@@ -186,3 +186,86 @@ func hasShards(confs queryrange.ShardingConfigs) bool {
 	}
 	return false
 }
+
+// NewSeriesQueryShardMiddleware creates a middleware which shards series queries.
+func NewSeriesQueryShardMiddleware(
+	logger log.Logger,
+	confs queryrange.ShardingConfigs,
+	middlewareMetrics *queryrange.InstrumentMiddlewareMetrics,
+	shardingMetrics *logql.ShardingMetrics,
+	limits queryrange.Limits,
+	merger queryrange.Merger,
+) queryrange.Middleware {
+
+	noshards := !hasShards(confs)
+
+	if noshards {
+		level.Warn(logger).Log(
+			"middleware", "QueryShard",
+			"msg", "no configuration with shard found",
+			"confs", fmt.Sprintf("%+v", confs),
+		)
+		return queryrange.PassthroughMiddleware
+	}
+	return queryrange.MiddlewareFunc(func(next queryrange.Handler) queryrange.Handler {
+		return queryrange.InstrumentMiddleware("sharding", middlewareMetrics).Wrap(
+			&seriesShardingHandler{
+				confs:   confs,
+				logger:  logger,
+				next:    next,
+				metrics: shardingMetrics,
+				limits:  limits,
+				merger:  merger,
+			},
+		)
+	})
+}
+
+type seriesShardingHandler struct {
+	confs   queryrange.ShardingConfigs
+	logger  log.Logger
+	next    queryrange.Handler
+	metrics *logql.ShardingMetrics
+	limits  queryrange.Limits
+	merger  queryrange.Merger
+}
+
+func (ss *seriesShardingHandler) Do(ctx context.Context, r queryrange.Request) (queryrange.Response, error) {
+	conf, err := ss.confs.GetConf(r)
+	// cannot shard with this timerange
+	if err != nil {
+		level.Warn(ss.logger).Log("err", err.Error(), "msg", "skipped sharding for request")
+		return ss.next.Do(ctx, r)
+	}
+
+	if conf.RowShards <= 1 {
+		return ss.next.Do(ctx, r)
+	}
+
+	req, ok := r.(*LokiSeriesRequest)
+	if !ok {
+		return nil, fmt.Errorf("expected *LokiSeriesRequest, got (%T)", r)
+	}
+
+	ss.metrics.Shards.WithLabelValues("series").Inc()
+	ss.metrics.ShardFactor.Observe(float64(conf.RowShards))
+
+	requests := make([]queryrange.Request, 0, conf.RowShards)
+	for i := 0; i < int(conf.RowShards); i++ {
+		shardedRequest := *req
+		shardedRequest.Shards = []string{astmapper.ShardAnnotation{
+			Shard: i,
+			Of:    int(conf.RowShards),
+		}.String()}
+		requests = append(requests, &shardedRequest)
+	}
+	requestResponses, err := queryrange.DoRequests(ctx, ss.next, requests, ss.limits)
```

Comment:
I guess this is where some of the magic happens. Multiple requests are made. Are they going to the same endpoint? How are they load balanced? Also, how is the endpoint defined?

Reply:
These are dispatched from the query frontend to queriers according to the sharding factor. The queriers then will hit the store (and ingesters if necessary - determined by the time range of the request since ingesters hold recent data). Both the store and the ingesters now have the ability to discern and filter results down to a specified shard.

Reply:
Ok. How are the queries routed to the inhesters?

Reply:
frontend -> queriers -> ingester
                                      -> hard storage.

Reply:
And if we have multiple queriers? Is the a load balancer?